### ISSUE TITLE
BUG: tf.random.normal() has a fixed value in eager mode (TF2.0)

### ISSUE THREAD

**Issue by markemus** [tensorflow/tensorflow#36715]
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1. It must be a bug, a feature request, or a significant problem with documentation (for small docs fixes please send a PR instead).
2. The form below must be filled out.
3. It shouldn't be a TensorBoard issue. Those go [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: 
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
- **TensorFlow installed from (source or binary)**:
- **TensorFlow version (use command below)**: v2.0.0-rc2-26-g64c3d38 2.0.0
- **Python version**: 3.6.9
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with:

```bash
python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"
```

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.

In Tf2.0 eager mode, tf.random.normal() will give the same value over and over again. This happens whether you use a Keras Model or just call the tf.random.normal() tensor repeatedly:

```
import numpy as np
import tensorflow as tf

id = np.ones(shape=(32,10))
i = tf.keras.layers.Input(shape=(10,), batch_size=32, dtype=tf.float64)
y = tf.random.normal(shape=(32,10), name="noise", dtype=tf.float64)
o = tf.add(i, y)
model = tf.keras.Model(inputs=i, outputs=o)

# same value every time?
model.predict(id)
model.predict(id)
model.predict(id)
```

This occurs without Keras as well:

```
x = tf.constant(value=np.ones(shape=(32,10)), dtype=tf.float64)
y = tf.random.normal(shape=(32,10), name="noise", dtype=tf.float64)
z = tf.add(x, y)
print(z)
print(z)
print(z)
```

If you disable eager mode with tf.compat.v1.disable_eager_execution(), the Keras Model will generate new values each time it's called (as it should).

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue. Please find the Gist [here](https://colab.sandbox.google.com/gist/amahendrakar/b5dfe95b4ce08d000c0cef3b05bb66d4/36715_eager.ipynb). Thanks!

**Comment by ymodak** [comment]
Variables state is stored in when you execute script. So if you are calling same variable multiple times you will see the same result.
However if you run the same script multiple times you can see different results.
Try executing following script multiple times (#avoid calling same variable multiple times in same script)
```python
import numpy as np
import tensorflow as tf
id = np.ones(shape=(2,2))
i = tf.keras.layers.Input(shape=(2,),dtype=tf.float64)
y = tf.random.normal(shape=(2,2), name="noise", dtype=tf.float64)
o = tf.add(i, y)
model = tf.keras.Model(inputs=i, outputs=o)
model.predict(id)
```
Similarly try;
```python
import numpy as np
import tensorflow as tf
x = tf.constant(value=np.ones(shape=(2,2)), dtype=tf.float64)
y = tf.random.normal(shape=(2,2), name="noise", dtype=tf.float64)
z = tf.add(x, y)
print(z)
```

**Comment by markemus** [comment]
This is not related to the seed- the issue is that the same code behaves differently in graph vs eager mode. The same code run in graph mode will return new values each time the tensor is evaluated, but in eager mode it stores those values and returns them repeatedly. 

This means that for example everyone using VAEs built in TF1.0 has code that is silently broken in TF2.0 (now that eager mode is the default)- instead of random values on each training step they're getting a fixed matrix that is instantiated once and evaluated over and over. 

Additionally there is no way as far as I can tell to get a tensor with new random values on each evaluation in eager mode. 

This same bug actually does exist in eager mode in TF1.0.

**Comment by foxik** [comment]
Well, to me the behaviour is exactly what I expect of the eager mode. When you have
```python
import tensorflow as tf
y = tf.random.normal(shape=(2,2), name="noise", dtype=tf.float64)
print(y)
print(y)
```
it should write the same value twice -- it was generated once during `tf.random.normal` call. So in your second example, I would expect three same values.

Similarly in your Keras code, `y` is generated once and never recomputed -- why should it? It is not a function generating random values, it was generated once and then stored in the model. If you want to generate fresh random values, you must change `y` to be a function of the input. For example, you could write
```python
y = tf.keras.layers.Lambda(lambda _: tf.random.normal(shape=(32,10), name="noise", dtype=tf.float64), dtype=tf.float64)(i)
```
I assume there is a better way to do it in Keras though.

**Comment by markemus** [comment]
It's extremely unexpected for a stochastic tensor to become non-stochastic in eager mode. It breaks working code without warning the user, and it does it in a very nasty hard to detect way. Your output will still be normally distributed. It will still change on every initialization (if you are not using a seed). I am 100% sure there are people out there atm working on VAEs that are actually AEs.

For people with working VAEs as well as beginners this is highly misleading, not least because [every](https://github.com/hwalsuklee/tensorflow-mnist-VAE/blob/master/vae.py#L74) [existing](https://github.com/y0ast/VAE-TensorFlow/blob/master/main.py#L49) [variational](https://github.com/ChengBinJin/VAE-Tensorflow/blob/master/src/vae.py#L51) [autoencoder](https://github.com/hardmaru/cppn-gan-vae-tensorflow/blob/master/model.py#L94) uses `tf.random_normal()` [now `tf.random.normal()` in TF2.0] directly to sample from a normal distribution. 

Using a lambda layer may work as a workaround but it doesn't solve the problem that this tensor silently changes its behavior between modes, and it shouldn't.

**Comment by foxik** [comment]
I do not think this is a silent change. In eager mode, `tf.random.normal()` is a function returning a TF tensor. When executed, it generated random number, and you get back a piece of memory containing these numbers -- there is no magic for the tensor to be regenerated each time you use it (which is how it worked in a graph mode, where `tf.random_normal()` returned a node in a graph, which was reevaluated every time you ran the graph).

Similarly, for `y = np.random.normal()`, you would not expect `y` to change.

**Comment by pandrey-fr** [comment]
I agree with @foxik, the current behaviour is strictly logical.
As to the proper way to add noise in TF2, as far as I know, it is by using a dedicated layer, _e.g._ the [GaussianNoise](https://www.tensorflow.org/api_docs/python/tf/keras/layers/GaussianNoise) one.

**Comment by markemus** [comment]
I hear what you guys are saying. But I think this is a big change from the way things worked in graph mode with no warning and no easy way to see the change. All four of the examples I linked used tf.random_normal the same way I did in the example- none of them wrap it in a layer.call. This has been the standard way to create VAEs for years now- all those people were following the same tutorials and if they've updated to TF 2.0, they've all got broken code now and don't know it. They could be publishing papers with that code and will need to issue retractions. 

If this change (which is not a change, but is nevertheless a change) is intentional it has _got_ to be documented, and imo there should be a warning when running in eager mode. There should also be a proper way to repeatedly sample from a normal distribution, because right now in eager mode there is no way to do this very common operation other than the workarounds you've posted (thanks for those btw).

**Comment by pandrey-fr** [comment]
> If this change (which is not a change, but is nevertheless a change) is intentional it has got to be documented, and imo there should be a warning when running in eager mode.

In my humble opinion, raising a warning whenever using `tf.random.normal` in eager mode would do more harm than good in that it would be over-triggered; yet I hear your point about the risk for incorrect code being ignorantly run, and I hope someone in the TF team will come up with a reasonable way to inform users. Perhaps tf1-to-tf2 conversion tools and tutorials already include checks on these cases, and if not it may be worth to enhance them in that direction?

> There should also be a proper way to repeatedly sample from a normal distribution, because right now in eager mode there is no way to do this very common operation other than the workarounds you've posted (thanks for those btw).

The simplest way is to run `tf.random.normal` iteratively, _i.e._ each time you want to use a new value. For example, something in the vein of the following would work:
```python
for sample in my_dataset:
    noise = tf.random.normal()
    output = model(sample + noise)
```
And when you want this operation to happen within a model, you similarly sample within the `call` method of a layer, because this is the code that is executed each time a sample is being processed. In TF1, the syntax is different, but the idea is similar : on each run, a symbolic tensor is given a random value sampled following the same generation procedure.

**Comment by markemus** [comment]
For posterity, here's a minimal example of how to correctly run the code in the OP with a custom Keras layer:

```
import numpy as np
import tensorflow as tf

class Sampler(tf.keras.layers.Layer):
    def __init__(self):
        super().__init__()
    def call(self, inputs):
        y = tf.random.normal(shape=inputs.shape, name="noise", dtype=inputs.dtype)
        z = tf.add(inputs, y)
        return z

# with Model
id = np.ones(shape=(32,10))
i = tf.keras.layers.Input(shape=(10,), batch_size=32, dtype=tf.float64)
# y = tf.random.normal(shape=(10,), name="noise", dtype=tf.float64)
# o = tf.add(i, y)
o = Sampler()(i)
model = tf.keras.Model(inputs=i, outputs=o)

# random is random now
model.predict(id)
model.predict(id)
model.predict(id)
```

That wasn't too bad. Alright, you guys convinced me- put something in the documentation and the update scripts (and maybe on the [migration page](https://www.tensorflow.org/guide/migrate) ) and I'll go away :)

**Comment by wangpengmit** [comment]
@foxik's two comments are exactly right, and I also agree with @markemus that this is a big change. Unfortunately `tf.random.normal` is very broken in TF2 (its semantics was designed for graph mode, and its statefulness makes it very hard to behave the same in TF1 and TF2). Please see https://www.tensorflow.org/guide/random_numbers for the recommended ways to generate random numbers TF2. We haven't migrated Keras to the new RNGs so we can't deprecate the old RNGs yet.

**Comment by pandrey-fr** [comment]
@wangpengmit Thank you for pointing the guide out. The (new ?) `Generator` class and its associated functions appear to still be in `tf.random.experimental` as of TF 2.1.0, something that the guide does not mention; to be clear, do you advise users to move to it now, or is it better (from a user standpoint) to wait for a future release to migrate home-written code?

**Comment by wangpengmit** [comment]
@pandrey-fr It will be `tf.random.Generator` in TF 2.2. I strongly encourage you to use it if you can. Some reasons you can't use it: (1) you need to create a `Generator` in a Distribution Strategy scope which is not allowed (we are working on allowing it, and there are workarounds as described in the guide); (2) you use Keras which doesn't use `Generator` yet.

**Comment by pandrey-fr** [comment]
Great, thank you for your the clarification!

**Comment by amahendrakar** [comment]
@markemus,
Could you please check @wangpengmit's comment above and let us know if your query is resolved. Thanks!

**Comment by markemus** [comment]
@amahendrakar the Generator class does seem like it a good replacement that would solve this issue. I've been using it myself per @wangpengmit 's suggestion. However I don't think this is resolved until tf.random.normal() is deprecated or there is a clear way for people to understand its different behavior in the different tensorflow modes. I'm still encountering subtly broken code in the wild.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.3 and the latest TF-nightly i.e. v2.4.0-dev20200805. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/dae30f344fcb72eff2140acab5270e7c/36715-2-3.ipynb). Thanks!

**Comment by sachinprasadhs** [comment]
Was able to reproduce your issue in Tf Nightly 2.6.0-dev20210524, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/34aa4b4ef47e4a09ebc3edfcf809a5cf/35650.ipynb). Thanks!

**Comment by sachinprasadhs** [comment]
There is already a warning message mentioned in the Random number generation document [here](https://www.tensorflow.org/guide/random_numbers). Which says `Warning: The old RNGs from TF 1.x such as tf.random.uniform and tf.random.normal are not yet deprecated but strongly discouraged.` 
Hope this clarifies your issue.

**Comment by markemus** [comment]
@amahendrakar @sachinprasadhs @wangpengmit  I kind of forgot about this issue, but it's still a problem- I actually saw tf.random.normal() being used in eager mode in a friend's project this week. 

My suggestion would be to either deprecate these old functions in tf.random, or add a warning that displays if these functions are called in eager mode, or at the very least ensure that there is a big red warning on the [documentation page](https://www.tensorflow.org/api_docs/python/tf/random/normal) for the function (currently there is nothing even indicating hat behavior differs in eager mode, let alone a warning).  

Could we please reopen this issue until we at least reach a state where the faulty behavior of these functions is clear to users? I do not seem to be able to reopen it myself. Thanks!

**Comment by markemus** [comment]
I found another way to get different behavior with tf.random.normal in eager vs graph mode. All you need is graph mode (with tf.compat.v1.disable_eager_execution), an exposed tf.random.normal() call (not wrapped in a function), a Keras model, and a tf.function to wrap the model, like so:

```

import numpy as np
import tensorflow as tf

# Try with both graph and eager mode
run_in_graph_mode = True
if run_in_graph_mode:
    tf.compat.v1.disable_eager_execution()
print(tf.executing_eagerly())

# Input is all ones, so the return value will just be y.
id = np.ones(shape=(3, 10))
i = tf.keras.layers.Input(shape=(10,), batch_size=3, dtype=tf.float64)
y = tf.random.normal(shape=(3, 10), name="noise", dtype=tf.float64)
o = tf.add(i, y)
model = tf.keras.Model(inputs=i, outputs=o)
# Wrap our model in a tf.function
f_model = tf.function(model)

print(model.predict(id))
print(model.predict(id))
print(model.predict(id))

x1 = f_model(id)
x2 = f_model(id)
with tf.compat.v1.Session() as sess:
    print(sess.run(x1))
    print(sess.run(x2))

```

Results:
eager mode: both model and f_model return fixed generated values.
graph mode: divergence, model returns unique generated values but f_model returns fixed generated values.

So far I have not found a way to make this bug occur just in TF2

**Comment by sachinprasadhs** [comment]
Do you find the below comment useful for you.
> There is already a warning message mentioned in the Random number generation document [here](https://www.tensorflow.org/guide/random_numbers). Which says `Warning: The old RNGs from TF 1.x such as tf.random.uniform and tf.random.normal are not yet deprecated but strongly discouraged.` Hope this clarifies your issue.

**Comment by markemus** [comment]
Well, the question is whether it affects graph mode in tf2 or only when using tf.compat.v1. So far I haven't been able to replicate the behavior with tf.Functions (ie tf2 graph mode) alone.

**Comment by sachinprasadhs** [comment]
In Tensorflow 2, if the code is working as intended then you can migrate your code to Tensorflow 2 compatible code, see guide here https://www.tensorflow.org/guide/migrate.
If the bug exists in Tensorflow 1 behavior, then it's most unlikely to get a fix for that version. Suggested way would be to use the latest stable Tensorflow version.

**Comment by markemus** [comment]
@sachinprasadhs  Okay, looks like the weird behavior does only occur in tf.compat.v1.disable_eager_execution() is enabled. tf..Function() does not have this issue. Thanks and feel free to close :)
